Zabbix agent registration from itself, added template checking and ad…

mmoono opened the pull request:
### What does this PR do?
It adds the possibility of running Zabbix API calls from either Zabbix Server or Zabbix Agent. It gives us the possibility of using grains['localhost'] as hostname/interface, when ran from agent. It also adds the possibility of assign newly added/existing host to a predefined template. 

### What issues does this PR fix or reference?
Currently you can only run Zabbix agent registration from the Zabbix master server. Also, when adding a new host to monitor, there is no way to assign it to existing template.

### Previous Behavior
zabbix_host.present state could be ran only from the zabbix server. New hosts were not assigned to a template. 

### New Behavior
zabbix_host.present can be ran from either zabbix_server or zabbix_agent. 
When specifying templateids in sls file, host is assigned to those templates. When template not specified, host will be added unassigned to any template.

### Tests written?

No

Commit:
Add network module for NILinuxRT: nilrt_ip.py

Connman is the network manager used as backend.
Functionalities:
  - ip.get_interface_details()
  - ip.up(interface)
  - ip.down(interface)
  - ip.set_dhcp_linklocal_all(interface)
  - ip.set_static_all(interface, address, netmask, gateway, domains)
  - ip.build_interface(interface, interface_type, enable, **settings)
  - ip.build_network_settings(**settings)
  - ip.get_network_settings()
  - ip.apply_network_settings(**settings)

Integration test for:
  - ip.enable
  - ip.disable
  - ip.set_dhcp_linklocal_all
  - ip.set_static_all

Signed-off-by: Heghedus Razvan <razvan.heghedus@ni.com>

onorua committed:
fix proposal for #34438

wildcard '*' support for pkg version comparison

terminalmage committed:
Better wildcard check

terminalmage committed:
yumpkg.py: add wildcard support to package installs

This also makes significant changes to the list_repo_pkgs function to
bring it in line with behavior that will be implemented for other pkg
providers.

terminalmage committed:
Add salt.utils.fnmatch_multiple

terminalmage committed:
aptpkg: add pkg.list_repo_pkgs function

terminalmage committed:
pacman: add support for wildcards in package installs

terminalmage committed:
Document wildcard versions in pkg.installed docstring

terminalmage committed:
Add wildcard versions to Nitrogen release notes

terminalmage committed:
Add integration test for wildcard versions

szviagintsev committed:
Simplify string manipulation

By explicitly checking the length of 'exc_str_prefix', omit the
try-except block, handling of the leading whitespace, and the comment on
what's going on.

szviagintsev committed:
Remove bad comment

The comment seems to stay there by mistake as it has no relation to
what's really going on (i.e. checking whether the device is already an
LVM PV). Remove it.

terminalmage committed:
Decrypt pillar override before it is passed to Pillar object

This allows pillar override to be available in an unencrypted state
during pillar compilation instead of clumsily bolting it on after the
fact.

terminalmage committed:
Add new config params for pillar decryption

terminalmage committed:
salt.utils: add default value for "default" in traversal funcs

harisokanovic committed:
system: Update hardware clock on date/time change

Update hardware clock so that date/time changes persist
through reboot.

Add hwclock check to system integration test. Bump tolerance
of _same_times() helper to account for longer run-time of set
operations. Updating hardware rtc's can take more than 2 seconds.

Signed-off-by: Haris Okanovic <haris.okanovic@ni.com>
Reviewed-by: Sergey Kizunov <sergey.kizunov@ni.com>
Reviewed-by: Alejandro del Castillo <alejandro.delcastillo@ni.com>
Reviewed-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

xbglowx committed:
dockerng support for --device

https://github.com/saltstack/salt/issues/33971

terminalmage committed:
Add a new config option for list of available pillar decryption renderers

Also rename the option which specifies the default, to reduce confusion.

terminalmage committed:
Rework minion-side pillar decryption for state subsystem

terminalmage committed:
Add pillar_enc support to CLI pillar in pillar.items

terminalmage committed:
Add salt.utils.crypt

terminalmage committed:
Decrypt pillar data at end of pillar compilation

terminalmage committed:
Update master config template with pillar decryption args

terminalmage committed:
Add documentation on pillar decryption config items

terminalmage committed:
Document pillar decryption feature in pillar docs

terminalmage committed:
Add Pillar Decryption info to Nitrogen release notes

carlpett committed:
Merge carlpett/salt-vault

terminalmage committed:
Add tests for pillar decryption

Commit:
Skip new managed file state test on Windows

Commit:
Update autodoc topics for new modules added in develop

Also adjusted versionadded tags as necessary on the files that
are getting new autodoc files added.

szviagintsev committed:
Grammar fixes in docstrings

Remove unneeded 'an' articles

cro committed:
Add CLI Examples so tests will pass

twangboy committed:
Fix geteuid stacktrace on Windows

twangboy committed:
Merge pull request #3 from twangboy/win_jenkins

Fix geteuid stacktrace on Windows

Commit:
[develop] Update release numbers for doc build (#39041)

Commit:
updated boto_lambda execution module with 'lambda environment variables'

vnitinv committed:
Merge pull request #5 from nembery/develop

Use salt template system for multiple backend support, avoid commit with no diff

chotti-nallo committed:
Bug fixes in facts_refresh and rpc

chotti-nallo committed:
Replace rpc-reply with rpc_reply in rpc function

vnitinv committed:
Merge pull request #6 from rajvidhimar/fix_facts_refresh

Bug fixes in facts_refresh and rpc

Commit:
linux_ip: Disable linux_ip for NILinuxRT systems

Signed-off-by: Heghedus Razvan <razvan.heghedus@ni.com>

The-Loeki committed:
Introduction of a new Salt-SSH cache roster with vastly expanded options

The-Loeki committed:
Salt-SSH cache roster lint fixes

chotti-nallo committed:
Changes in facts dictionary

chotti-nallo committed:
proxy

chotti-nallo committed:
Change grains

Commit:
Merge branch '2016.11' into 'develop'

Conflicts:
  - salt/modules/vsphere.py

vnitinv committed:
Merge pull request #7 from rajvidhimar/fix_facts_refresh

Fix facts refresh

Commit:
Merge pull request #39064 from rallytime/merge-develop

[develop] Merge forward from 2016.11 to develop

Inveracity committed:
bugfix adding subnetwork option to gce

Previously when deploying to custom networks salt-cloud would fail with the error described in #31071
Now it is possible to specify a subnetwork in the profile, to deploy to a custom network and a specific subnet in that network.

szviagintsev committed:
Fix test name (#39061)

Remove an unnecessary underscore character from the name of the
linux_lvm.lvdisplay test:

  test__lvdisplay -> test_lvdisplay

Commit:
Fix some spelling errors and remove unecessary doc modules

gtmanfred committed:
use_keystoneauth should default to False if not specified

Commit:
Merge pull request #39074 from sergeizv/remove-bad-comment

Remove bad comment

Commit:
Merge pull request #39073 from sergeizv/grammar-fixes-in-docstrings

Grammar fixes in docstrings

Commit:
Additional newline to separate funcs

Needed for pylint

Commit:
function_present for boto_lambda state module now accepts Lambda Environment Variables.

Commit:
update botocore dependencies, skip if not met.

Commit:
Merge pull request #39043 from twangboy/develop

SaltTesting fixes for Windows

Commit:
Fix trailing whitespace

Commit:
Merge pull request #39079 from gtmanfred/2016.3

use_keystoneauth should default to False if not specified

Commit:
Merge pull request #39080 from kraney/boto_lambda_envvar

Added Lambda environment variables support to boto_lambda modules/states

Commit:
Merge pull request #39072 from Inveracity/patch-3

bugfix adding subnetwork option to gce

Commit:
Merge pull request #39031 from cloudflare/GRAINS-RUNNER

mine.update runner

Commit:
Rework how the loader pulls in minion opts in namecheap utils file

bmwiedemann committed:
Add salt-minion.sleep

for use with systemd to be installed in
/usr/lib/systemd/system-sleep/salt-minion.sleep

To avoid stale minion-master connections as described in
https://bugzilla.opensuse.org/show_bug.cgi?id=1018791

Commit:
Merge pull request #39002 from rallytime/fix-38111

Pass win_* options as kwargs when calling file.manage_file

Commit:
Merge pull request #39007 from cloudflare/SALT-EXEC

New runner function: salt.execute

The-Loeki committed:
Remove commented-out debug lines

terminalmage committed:
Run test_valid_docs in batches

This test fails often due to being trimmed, thanks to
salt.utils.dicttrim trimming values > max_event_size.

This commit changes this test so that it runs sys.doc twice, ensuring
that the return from sys.doc (for now) is not trimmed.

Commit:
add ability to override pillar data during state.sls_id run.

terminalmage committed:
Account for trimmed value in 'salt -d' output

terminalmage committed:
Don't fail test if data is empty

This keeps trimmed output from failing the test. We are still testing
with a specific module lower down in the test, so this doesn't reduce
our test coverage.

Commit:
Add a function to list extmods

Commit:
Doc example

terminalmage committed:
add debug logging for batch vars

mirceaulinic committed:
Log module

githubcdr committed:
Do not use name resolving for --notrim check

This causes a significant delay on servers with many open connections

chotti-nallo committed:
Fix documentation

vnitinv committed:
Merge pull request #8 from rajvidhimar/fix_facts_refresh

Fix documentation

vnitinv committed:
Merge remote-tracking branch 'upstream/develop' into develop

chotti-nallo committed:
Resovle conflicts

terminalmage committed:
make helper function private

mcalmer committed:
docker buildargs require minimal dockerpy 1.6.0 and docker version 1.9.0

mcalmer committed:
update version in the docs

mirceaulinic committed:
BGP runner

mirceaulinic committed:
BGP runner doc

Commit:
Merge pull request #39104 from githubcdr/develop

Do not use name resolving for --notrim check

Commit:
Merge pull request #39103 from arif-ali/fix_iface_lower

fixes #31393 don't convert the iface to lower case

Commit:
Merge pull request #39082 from bmwiedemann/systemdsleep

Add salt-minion.sleep

Commit:
Merge pull request #39060 from harisokanovic/dev/hokanovi/hwclock-sync

system: Update hardware clock on date/time change

Commit:
Merge pull request #39050 from The-Loeki/cache-roster2

Introduction of a new Salt-SSH cache roster with vastly expanded options

Commit:
Merge pull request #39036 from cro/no_docs_develop

Add CLI Examples so tests will pass (counterpart to #39035)

Commit:
use the pillar kwarg explicitely

mirceaulinic committed:
Jinja topic doc: logs subsection

Commit:
Merge pull request #39114 from cloudflare/BGP-RUNNER

New networking runner: BGP

mirceaulinic committed:
do instead of set

garethgreenaway committed:
Fixing a weird edge case when using salt syndics and targetting via pillar.  Without this fix the master of masters ends up in an infinite loop since the data returned from the minions is differently structured than if a sync was not in use.

garethgreenaway committed:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch '2016.11' into 'develop'

Conflicts:
  - doc/ref/runners/all/index.rst
  - salt/modules/win_lgpo.py
  - salt/states/dockerio.py

eradman committed:
Allow pillar environment to be specified for salt-run

Example:
  salt-run state.sls deploy saltenv=base pillarenv=dev

chotti-nallo committed:
Enhance install_config and install_os

10ne1 committed:
grains/core: Populate hw info for ARM devices (#38992)

Linux ARM targets use U-BOOT / fw_printenv and have no SMBIOS /
dmidecode installed. Set the manufacturer and serial number grains
using fw_printenv.

Signed-off-by: Ioan-Adrian Ratiu <adrian.ratiu@ni.com>

garethgreenaway committed:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge pull request #39129 from rallytime/merge-develop

[develop] Merge forward from 2016.11 to develop

garethgreenaway committed:
Missing colon

mirceaulinic committed:
Remove CLI restriction

cro committed:
MANY dunder variable fixes for proxies + proxy keepalive from @mirceaulinic (#38829)

* Add proxy keep alive options

* Documentation for the alive function

* MANY dunder variable fixes for proxies.

* Lint.

* Lint.

* More Lint.

Commit:
Merge pull request #39130 from garethgreenaway/develop

Fix when targeting via pillar with Salt syndic

Commit:
Merge pull request #39092 from jacobweinstock/add-pillar-data-override-in-state-sls-id

add ability to override pillar data during state.sls_id run.

Commit:
Fix six import

Commit:
Avoid reserved kw

Commit:
Merge pull request #39004 from markuskramerIgitt/compile_offline3

Compile offline3

Commit:
Merge pull request #39027 from cloudflare/REDIS-CACHE

Redis cache plugin

Commit:
Accidentially overwrote lint commit

Commit:
Merge pull request #39093 from cachedout/zd_1151

Add a function to list extmods

a-schuurman committed:
Added ipv4 arg to connect_container_to_network

Commit:
[PY3] Test comment contents more carefully for list stability

Stabilizes these tests by avoiding errors when cheese, crackers
order changes in the comment and gives errors like:

-  'comment': 'Kernel modules crackers, cheese are already present',
?                                     --------

+  'comment': 'Kernel modules cheese, crackers are already present',
?                            ++++++++

Commit:
[PY3] Sort return data and mock data for testing stability

mmoono committed:
Zabbix agent registration from itself, added template checking and adding

mmoono committed:
Zabbix agent registration from itself, added template checking and adding

chotti-nallo committed:
Update rpc to display reply in different format as well as update docstring

nobourbakist committed:
[Bugfix] fix argument order for usermod call

The argument order for the usermod -p call was reversed.

Commit:
Merge pull request #39160 from nobourbakist/fix-usermod-args

[Bugfix] fix argument order for usermod call

sjorge committed:
zone state for Solaris 10/illumos (#39155)

* create zone state, zone.running, and zone.stopped

* renamed zone.running and zone.stopped, minor lint fixes

* cleanup message in zonecfg to make it clear for the zone state

* zone.property_present state

* zonecfg.set_property - should handle true/false parsed as boolean

* zonecfg module should be more consistant with message

* zone.property_absent implemented

* fix more minor issues in zonecfg when values get parsed as bool

* zonecfg fix spelling error

* improved zonecfg error message handling

* zonecfg parser should also do the reverse

* more improvements in zonecfg for parsing values

* small tweak to zonecfg simplyfy code for the zone state

* Add resource_present to zone state

* zonecfg lint fixes

* resource_absent state, lint fixed, test support

* zonecfg should be a bit more careful when removing the exported config

* zone.export state

* zone.import state

* zone.absent implemented

* zone.attached and zone.detached implemented

* very minor consistency fixes

* added zone.installed and zone.uninstalled, made zone.detached and zone.halted not fail on missing zone

* handle bool values in _parse_value

* add useful cleanup internal function to  zonecfg

* zonecfg - cleaned up _parse_value, added _sanitize_value and _dump_cfg for debugging

* zone.absent - uninstall would sometimes not do the right thing. some work on zone.present

* zone.present implemented path for when config does not yet exist

* zone.present supports updating properties

* zonecfg should support resources that do not have selectors

* zonecfg should also debug log remove_resource its config

* zone.present - implemented, after 2 refactors

* Finish documentation

Commit:
Merge pull request #39150 from rallytime/python3-compat-fixes

Python 3 Test Fixes

Commit:
Merge pull request #39148 from a-schuurman/feature/dockerng-module-ipv4

Added ipv4 arg to connect_container_to_network

Commit:
Merge branch '2016.11' into 'develop'

No conflicts.

twangboy committed:
Add publisher_acl and publisher_acl_blacklist to defaults

Commit:
Fix doc build errors in junos_syslog engine

- Moved the module-level Parser() call to the class level
- Added twisted imports to list of MOCK_MODULES
- Also added msgpack to handle some other errors

Commit:
Added missing source_hash_name argument in get_managed function
Additional fix to  #33187
Customer was still seeing errors, this should now work.
Tested with 2015.8.13 and 2016.11.2

techhat committed:
Add __opts__ and __utils__ to SDB

techhat committed:
Type check for utils

twangboy committed:
Updates to win_iis modules

twangboy committed:
change to return_json

twangboy committed:
Remove raise from _srvmgr

bwesterb committed:
pip_state: mention Python 2 pip module is required

To use the pip state for Python 3, the pip module for Python 3 *and* Python 2 are required to be installed.  This adds "Python 2" in the error message when the Python 2 pip module can't be found.

thatnerdjosh committed:
Fix crashes when source attr is missing, as well as when description tag is missing in source image

Inveracity committed:
update GCE docs adding subnetwork parameter

Updating docs to reflect merge #39072

mchugh19 commented:
Just a head's up, there is template functionality in the state in the develop branch. See https://github.com/saltstack/salt/blob/develop/salt/states/zabbix_host.py

It handles adding a template via templateId or template Name. So you may wish to look into leveraging that as well.

mmoono commented:
Oh, thanks for the info Christian. Of course we will use it once it's available in a release.
What about launching zabbix commands from zabbix agent hosts?

Commit:
Merge pull request #39183 from Inveracity/patch-3

update GCE docs adding subnetwork parameter

Commit:
Merge pull request #39177 from bwesterb/patch-1

pip_state: mention Python 2 pip module is required

Commit:
Merge pull request #39176 from twangboy/win_iis

Updates to win_iis modules

Commit:
Merge pull request #39170 from grep4linux/jboss-state-fix

Added missing source_hash_name argument in get_managed function

Commit:
Merge pull request #39168 from twangboy/fix_default_config

Add publisher_acl and publisher_acl_blacklist to defaults

Commit:
Merge pull request #39167 from rallytime/merge-develop

[develop] Merge forward from 2016.11 to develop

Commit:
Merge pull request #39140 from eradman/pillarenv

Allow pillar environment to be specified for salt-run

Commit:
Merge pull request #39111 from mcalmer/update-min-version-dockerng

Update min required versions in dockerng

Commit:
Merge pull request #39099 from cloudflare/JINJA-LOG

Log module

carlpett committed:
Clarify where sdb profiles need to be configured

Commit:
Merge pull request #39034 from rallytime/doc-updates-develop

Update autodoc topics for new modules added in develop

Commit:
Indexes for format statement

ninja- committed:
Change reload_modules to trigger only on state changes (#39015) (#39016)

* Change reload_modules to trigger only on state changes (#39015)

* Helps with performance
* 99.9% of the time used by Salt users along with pip.installed, pkg.installed or cmd.run which remain compatible.
* Creates minor incompatibility if used with a logically-incomplete state that does not properly report changes.

* Add option to force module reload even with no changes at state.py

* fix force_reload_modules logic

Commit:
Merge pull request #39013 from terminalmage/zh878

Decryption of values in Pillar SLS files

spidercensus committed:
Update junos.py

Called .keys() once instead of 10 times.

Commit:
Merge pull request #38781 from raazvvann/dev/rheghedu/nilrt_ip

Add network module for NILinuxRT: nilrt_ip.py

cro committed:
Merge pull request #38339 from cloudflare/STRAIGHT-MINION

Migrate NAPALM modules to be used in both proxy and regular minion

marcaurele committed:
Doc: add getting started page for CloudStack provider based on Exoscale as an example

Signed-off-by: Marc-Aurèle Brothier <m@brothier.org>

thatnerdjosh committed:
Fix lint errors

Commit:
Revert "Cache the renderer in the pillar"

This reverts commit bd12f7e2698f23b387cbf71d3003893025aee99a.

twangboy committed:
Add start/stop/restart site, improve docs

xbglowx committed:
Remove unneeded log.error

twangboy committed:
Add start and stop apppool commands

thatnerdjosh committed:
Change == to is for comparison to None

vnitinv committed:
Merge pull request #4 from rajvidhimar/bug-fixes

Fix install_config

vnitinv committed:
Merge branch 'develop' into develop

vnitinv committed:
Merge pull request #9 from spidercensus/develop

Add optional args from jnpr.junos.Device to proxy module

vnitinv committed:
Merge remote-tracking branch 'upstream/develop' into develop

Mrten committed:
remove (?x) modifiers

at the start of the regex there can be modifiers which also have to be removed, see #39188

Commit:
Merge pull request #39207 from twangboy/win_iis

Additional functions to win_iis

Commit:
Merge pull request #39201 from marcaurele/doc/cloudstack

Doc: add getting started page for CloudStack provider

Commit:
Merge pull request #39196 from carlpett/patch-3

Clarify where sdb profiles need to be configured

Commit:
Merge pull request #39182 from NerdsvilleCEO/develop

Fix crashes when source attr is missing, as well as when description is missing

cachedout commented:
@mchugh19 Did you see the comment above from @mmoono ?

s0undt3ch committed:
Gate 3rd-party library imports (#39223)

Commit:
Merge pull request #39033 from xbglowx/dockerng_devices

dockerng support for --device

Commit:
Merge pull request #39205 from cachedout/revert_36525

Revert "Cache the renderer in the pillar"

techhat committed:
Pass in existing __utils__ instead of making a new one

mchugh19 commented on salt/modules/zabbix.py:
```diff
@@ -44,13 +44,12 @@
+    Check whether zabbix_server or zabbix_agent are installed.
     '''
-    if salt.utils.which('zabbix_server'):
+    if salt.utils.which('zabbix_server') or salt.utils.which('zabbix_agent'):
```
If this is to be opened up to machines other than the zabbix server, you don't need this at all. The rest of the module seems to use salt's http interfaces, so it should run without needing zabbix utilities installed.

mchugh19 commented on salt/modules/zabbix.py:
```diff
@@ -1452,3 +1456,109 @@ def hostinterface_update(interfaceid, **connection_args):
+        return False
+
+
+def template_get(templateids=None, groupids=None, hostids=None, **connection_args):
```
You want to submit this against the develop branch. This function already exists there, so this will need a conflict merge.

mchugh19 commented on salt/states/zabbix_host.py:
```diff
@@ -143,7 +144,25 @@ def _interface_format(interfaces_data):
             update_hostgroups = True
 
-        hostinterfaces = __salt__['zabbix.hostinterface_get'](hostids=hostid)
+        hosttemplates = __salt__['zabbix.template_get'](hostids=hostid, **kwargs)
```
The develop branch version of this fine includes another function, assign_templates. You may want to base this off of that functionality as it allows for specifying a template via a templateID or name.

mchugh19 reviewed (changes requested):
Will need a rebase against develop

mchugh19 commented:
@mmoono Yep, it sounds like a fine idea to me! In our environment we didn't want to push out zabbix admin credentials out to all minions. If you're okay with that, all should be fine. You'll want to rebase this work against the develop branch, as some of this functionality already exists there.

techhat committed:
Add import back in

mmoono commented:
@mchugh19, thanks! We didn't check the develop branch earlier to see what's there. As you said, the template handling is already there, so my changes are not needed. We will rebase the modules with develop branch tomorrow morning. But, again as you said ;), passing the connection arguments to the zabbix agents (in zabbix_host state) could come in handy.
Thanks!

Commit:
Fixup docs that don't have a CLI Example

- log functions shouldn't be used at the CLI level: added to allow_failure in test
- Added example for status.proxy_reconnect

Commit:
Merge pull request #38621 from terminalmage/pr34442

Add support for wildcard versions in pkg.installed states

Commit:
Merge pull request #39174 from techhat/sdbdunder

Add __opts__ and __utils__ to SDB

Commit:
Merge pull request #39229 from rallytime/doc-test-fix

Fixup docs that don't have a CLI Example

Commit:
Merge pull request #37808 from cloudflare/JINJA-FILTERS

Jinja filters

chotti-nallo committed:
Minor change in junos proxy to accept pillar named username. Few changes in junos module.

chotti-nallo committed:
Changes in junos proxy

Commit:
Fixes #39214. FTP authentication and clean disconnect.

vnitinv committed:
Merge pull request #10 from rajvidhimar/bug-fixes

Minor change in junos proxy to accept pillar named username. Few changes in junos module

marcaurele committed:
Add link in toc for cloudstack page

gtmanfred committed:
add runas as a global state variable

ClaudiuLSH committed:
Fixes icinga2 module return code (#38987)

* Fixes icinga2 module return code

* Remove unused os import

vnitinv committed:
Merge remote-tracking branch 'upstream/develop' into develop

freach committed:
Pull request for #39193 (#39244)

* improved log_config documentation for #39193

* added log_config documentation for #39193

Commit:
Merge pull request #39245 from marcaurele/doc/toc-for-cloudstack

Add link in toc for cloudstack page

Commit:
Merge pull request #39242 from adammichalik/salt-39214_ftp_authentication

Fixes #39214. FTP authentication and clean disconnect.

gtmanfred committed:
add optional extmod_whitelist dictionary

gtmanfred committed:
add documentation for the global runas state option

gtmanfred committed:
add extmod_whitelist to the saltutil modules

gtmanfred committed:
document extmod_whitelist

for master and minion configs

gtmanfred committed:
add sync_sdb to salt-run saltutil

gtmanfred committed:
add note to clean_dynamic_modules

KaiSforza committed:
Add nix module to Salt

This module adds the ability to work with nix packages, and do some
basic operations on the nix store. There is still some work to be done,
but this lets you orchestrate some basic tasks for a system with nix
installed.

gtmanfred committed:
add clean_dynamic_modules to master default opts

gtmanfred committed:
add integration test

Commit:
import re

Commit:
add boto_iam.export_roles, like boto_iam.export_users

Commit:
comments

Commit:
Fixing issue #39161, salt-ssh : match:grain for pillars inconsistencies

gtmanfred committed:
fix pylint

rajvidhimar committed:
Lint and other changes

Commit:
Refactor blockdev to use blkid instead of lsblk

szviagintsev committed:
Do boolean check instead len() to test whether string is empty

Do a truth check to test whether a string is empty instead of using the
len() function. As @terminalmage mentioned in #39056, this is about
twice faster; also this is what PEP8 says to do.

rajvidhimar committed:
Change indentation

vnitinv committed:
Merge pull request #11 from rajvidhimar/bug-fixes

Lint and other changes

mcalmer committed:
support dryrun for dockerng.sls_build

rajvidhimar committed:
Add safety mechanism in shutdown fucntion and add passwd as valid argument

rajvidhimar committed:
Remove blank line

vnitinv committed:
Merge pull request #12 from rajvidhimar/bug-fixes

Add safety mechanism in shutdown function and add passwd as valid arg…

Commit:
salt/states/blockdev.py

mcalmer committed:
provide the possibility to put extra modules into the thin

Commit:
Add __utils__ to sdb unit test (#39261)

Refs #39174

Commit:
Fix the new jinja_test failures (#39262)

Refs #37808

The max fuction returns an int, not a float, as does the min function.
This updates the assertions to compare against ints.

Commit:
Purged trailing whitespace

Commit:
Merge pull request #39265 from edgan/roster_grain_pillars

Fixing issue 39161, salt-ssh : match:grain for pillars inconsistencies

Commit:
salt.utils.gitfs: remove dulwich support, make refspecs configurable (#39210)

* Add config params for custom refspecs

* Add exceptions for errors encountered modifying git config

* Make the refspecs a configurable parameter

* Make refspecs a per-remote parameter

* Update master config template to include custom refspecs opts

* Add documentation for new config params

* Update GitFS walkthrough with a section on custom refspecs

* Remove dulwich support from salt.utils.gitfs

Dulwich still lacks important features, including (but not limited to)
the following:

- Lack of the necessary support for checking out a ref needed for
  git_pillar/winrepo support

- No support in its config objects for multivar git config items, making
  it impossible to detect when repos have multiple refspecs set for a
  given git remote

Given this information, and the fact that it trails as a distant third
to Pygit2 and GitPython, Salt will cease to support Dulwich as a git
interface moving forward.

* Excise references to dulwich from documentation

* Add mention of custom refspecs to Nitrogen release notes

* Add gitfs_refspecs to mocked opts in gitfs integration tests

Also remove dulwich from unit tests

* Add information about opts argument not being intended for CLI use

Commit:
Merge pull request #39263 from bsherrod/export_iam_roles

add boto_iam.export_roles execution module function

Commit:
Merge pull request #39252 from gtmanfred/extmod_whitelist

Allow specifying which modules should be used when running saltutil.sync_*

Commit:
Merge branch '2016.11' into 'develop'

Conflicts:
  - salt/config/__init__.py
  - salt/modules/win_lgpo.py
  - salt/utils/aws.py
  - tests/unit/utils/schema_test.py

Commit:
Merge pull request #39248 from gtmanfred/develop

add runas as a global state variable

Commit:
Remove unused imports

Commit:
Merge pull request #39056 from sergeizv/simplify-string-manipulation

Simplify string manipulation

Commit:
Merge pull request #39278 from mcalmer/dockerng-dryrun-for-sls_build

Dockerng dryrun for sls build

Commit:
Merge pull request #39272 from daxroc/fix/lsblk_to_blkid

Refactor blockdev to use blkid instead of lsblk - fixes #39255

s0undt3ch committed:
Use six.moves to import cStringIO

s0undt3ch committed:
Restore full user permissions when failing to delete

s0undt3ch committed:
Inform on which python version we are running on

s0undt3ch committed:
Merge pull request #39285 from s0undt3ch/features/py3

Features/py3

KaiSforza committed:
nix: Check for nix-collect-garbage

I changed this from nix-store to nix-collect-garbage as the
nix.collect_garbage function used to use `nix-store --gc`, but now just
uses nix-collect-garbage.

KaiSforza committed:
nix: Correct documentation for gc

It doesn't actually show a list of all the operations, just the nix
summary.

Commit:
Merge pull request #39250 from Juniper/develop

junos specific changes to add more feature and enhancements to our APIs

Commit:
Fix up bad merge

Commit:
Merge pull request #39267 from KaiSforza/nix

Add nix module to Salt

Commit:
Merge pull request #39283 from rallytime/merge-develop

[develop] Merge forward from 2016.11 to develop

Commit:
Merge pull request #39212 from carlpett/feature/vault-integration

Merge carlpett/salt-vault

Commit:
Merge pull request #38767 from github-abcde/improve-templating

template.py: Fix error thrown if default passed to compile_template i…

Commit:
Merge pull request #39215 from Mrten/patch-1

remove (?x) modifiers in file.comment regex

Commit:
Protect hardware clock test comparison better (#39298)

In #39060, some tests were added to check for the hardware clock
getting reset when possible. The test attempted to gate the test
when access to the hardware clock was impossible, but the check
wasn't quite right and caused problems with the resulting tests.

This change makes the `has_settable_hwclock()` function public
in order to make the test gate work correctly. This function
could also be useful at the CLI level.

When running `self.run_function('status._has_settable_hwclock()')`
from the test suite, the return is a string noting that the
function is unavailable. This change allows for the `if not` check
to evaluate the return of the function.

mcalmer committed:
added unit test for dockerng.sls_build dryrun option (#39305)

terminalmage committed:
Fix gpg renderer breakage from bef5e66

bef5e66 sought to make a PY3 compatibility change, but the attribute
from Python2's cStringIO module is not available via the "fake"
cStringIO which six exposes, as six's cStringIO is just the
cStringIO.StringIO function. The other cStringIO class attributes are
not available via six.

This commit adds a new module with a few functions to get information on
StringIO objects both in Python2 and 3, and modifies the files edited in
bef5e66 to use this new module to determine if the object is a readable
StringIO.

Fixes #39334.

s0undt3ch committed:
Merge pull request #39340 from terminalmage/issue39334

Fix gpg renderer breakage from bef5e66

mchugh19 commented:
Hey @mmoono! Hopefully you can get the rebase done, and evaluate the template handling. In looking things over, it turns out the kwargs/connection_args are often not passed to the module calls, making the comments about optional _connection_user/pass/url a complete lie :/

We've just started investigating using vault to access passwords. As part of that, we're looking at migrating password data from pillars to a query in the top of some states, which means using the _connection_whatever functionality. 

Cheers!

mmoono commented:
Hi @mchugh19, 
Indeed, I didn't pass the **kwargs to all __salt__ calls in the zabbix_host.py. Now, this is corrected. 
All in all, the methods which we were testing in our env, did pass the kwargs correctly. API calls were made from the minion to zabbix server without any problems. 
We will now wait for newer version of Salt, to take it into use.
Thanks!

Take care.

mchugh19 commented:
Super! Can you push the updates? One or two of the calls had overloaded the name kwarg, where passing in all **kwargs might make things funky. So we'll just want to review.

mmoono commented:
It should be there already: https://github.com/mmoono/salt/commit/8c2cdf77c9287682bc0d7a9ad3508321c8b24f93

mchugh19 commented:
Gocha. Didn't see that as it wasn't pushed to this PR.

cachedout commented:
Go Go Jenkins!

mmoono commented:
@mchugh19, damn! Sorry for that. Ehh, it looks like I have a lot to learn when it comes to git..

cachedout commented:
@mmoono Did you want to try to get that commit into this PR? Let me know if I can help. Thanks!

cachedout commented:
We also need the following two errors fixed before we can merge this, please: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/8539/violations/

mmoono commented:
@cachedout, I've fixed the pylint violations: https://github.com/mmoono/salt/commit/05385aba54b623ba38dfe90dda7dc286987e30f1. 
But now comes the tricky part(?). How can I add different branch (in this case develop) to this PR, while it is based on the 2016.11 branch?

cachedout commented:
@mmoono In order for those pylint (and other) fixes to show up in this PR, they made to be made and pushed to the same branch that this PR is made from. In this case, that's `mmoono:2016.11`

mmoono committed:
Adding develop changes

mmoono committed:
Fixing pylint

cachedout commented:
@mmoono We still don't have the right commits in this PR. I'm going to go ahead and close this and ask you to please open a brand-new PR with the changes that you want made. Thanks.
